src/net.rs

```diff
@@ -87,12 +87,14 @@ pub(crate) async fn connect_tcp(
                     context,
                     "Failed to connect to {}: {:#}.", resolved_addr, err
                 );
-                last_error = Some(err);
+                if first_error.is_none() {
```

Comment:
Afaiu this can be shortened to `first_error.get_or_insert(err);`, but didn't try